Sources/NIOFileSystem/Internal/ParallelDirCopy.swift

```diff
@@ -0,0 +1,158 @@
+            // - Updates activeDirCount and finishes the stream if required.
+            // - Either adds a single task on `taskGroup` or none.
+            // Returns true if it added a task, false otherwise.
+            func onNextItem(_ item: DirCopyItem) -> Bool {
+                switch item {
+                case .endOfDir:
+                    activeDirCount -= 1
+                    if activeDirCount == 0 {
+                        sequence.source.finish()
+                    }
+                    return false
+                case let .toCopy(from: from, to: to):
+                    if from.type == .directory {
+                        activeDirCount += 1
+                    }
+                    taskGroup.addTask {
+                        try await copySelfAndEnqueueChildren(
+                            from: from,
+                            to: to,
+                            yield: yield,
+                            shouldProceedAfterError: shouldProceedAfterError,
+                            shouldCopyItem: shouldCopyItem
+                        )
+                    }
+                    return true
+                }
+            }
+
+            let iter = sequence.sequence.makeAsyncIterator()
+
+            // inProgress counts the number of tasks we have added to the task group
+            // Get up to the maximum concurrency first.
+            // We haven't started monitoring for task completion, so inProgress is 'worst case'.
+            var inProgress = 0
+            while inProgress <= maxConcurrentOperations {
+                try Task.checkCancellation()
+                let item = await iter.next()
+                if let item = item {
+                    if onNextItem(item) {
+                        inProgress += 1
+                    }
+                } else {
+                    // We completed things before we hit the limit
+                    return
+                }
+            }
+
+            // Then operate one in (finish) -> one out (start another),
+            // but only for items that trigger a task.
+            while let _ = try await taskGroup.next() {
+                var keepConsuming = true
+                while keepConsuming {
+                    try Task.checkCancellation()
+                    let item = await iter.next()
+                    if let item = item {
+                        keepConsuming = !onNextItem(item)
+                    } else {
+                        keepConsuming = false
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// An 'always ask for more' no  back-pressure strategy for a ``NIOAsyncSequenceProducer``.
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+private struct NoBackPressureStrategy: NIOAsyncSequenceProducerBackPressureStrategy {
+    public mutating func didYield(bufferDepth: Int) -> Bool { true }
+
+    public mutating func didConsume(bufferDepth: Int) -> Bool { true }
```

Comment:
nit: `public` is redundant and misleading here

```suggestion
    mutating func didYield(bufferDepth: Int) -> Bool { true }

    mutating func didConsume(bufferDepth: Int) -> Bool { true }
```